go/arrow/cdata/cdata.go

```diff
@@ -418,26 +409,28 @@ func (imp *cimporter) doImportChildren() error {
 
 func (imp *cimporter) initarr() {
 	imp.arr = C.get_arr()
+	if imp.alloc == nil {
+		imp.alloc = &importAllocator{arr: imp.arr}
+	}
+}
+
+func (imp *cimporter) doImportArr(src *CArrowArray) error {
+	imp.arr = C.get_arr()
+	C.ArrowArrayMove(src, imp.arr)
+	if imp.alloc == nil {
+		imp.alloc = &importAllocator{arr: imp.arr}
+	}
+	return imp.doImport()
 }
 
 // import is called recursively as needed for importing an array and its children
 // in order to generate array.Data objects
-func (imp *cimporter) doImport(src *CArrowArray) error {
-	imp.initarr()
+func (imp *cimporter) doImport() error {
 	// move the array from the src object passed in to the one referenced by
 	// this importer. That way we can set up a finalizer on the created
 	// arrow.ArrayData object so we clean up our Array's memory when garbage collected.
-	C.ArrowArrayMove(src, imp.arr)
 	defer func(arr *CArrowArray) {
-		if imp.data != nil {
-			runtime.SetFinalizer(imp.data, func(arrow.ArrayData) {
-				defer C.free(unsafe.Pointer(arr))
-				C.ArrowArrayRelease(arr)
-				if C.ArrowArrayIsReleased(arr) != 1 {
-					panic("did not release C mem")
-				}
-			})
```

Comment:
Might be a good idea to keep the finalizer in debug mode to assert that the ref count that's maintained manually via Retain/Release is indeed zero when the objected is not referenced anymore.

Reply:
We don't actually expose any API to externally check the current refcount on the objects so there's not really a way for the finalizer to check that. That's why I added the new tests to confirm that `Release` does actually free the memory, and users can use the `CheckedAllocator` if they want to ensure that things are being properly released

Reply:
Ahh. Makes sense. I do remember unit tests complaining about my refcounts when I was implementing ListView.